### PR TITLE
Initial sample for ServiceHub

### PR DESCRIPTION
**Pull request by kzu**

### CODE REVIEW

File: ClientServerApp/ServiceHub.Echo/EchoService.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            // own brokered service, by using the IServiceBroker we got in the constructor from 
+            // out IServiceHubServiceFactory implementation.
+            if (int.TryParse(message, out var number))
+            {
+                var encoder = await broker.GetProxyAsync<IEncodingService>(
+                    ServiceDescriptors.EncodingService, cancellationToken: cancellation);
+
+                // Showcases how to properly dispose the proxy, even if the underlying service interface, 
+                // IEncodingService in this case, doesn't implement IDisposable
+                using (encoder as IDisposable)
```
Comment (RyanToth3) [inline ClientServerApp/ServiceHub.Echo/EchoService.cs:44]: I would null check the encoder here

File: README.md [hunk 2]
```diff
@@ -1,53 +1,29 @@
...
-
-## Setup
-
-Explain how to prepare the sample once the user clones or downloads the repository. The section should outline every step necessary to install dependencies and set up any settings (for example, API keys and output folders).
-
-## Running the sample
+Latest Visual Studio (2019+) with the following workloads:
 
-Outline step-by-step instructions to execute the sample and see its output. Include steps for executing the sample from the IDE, starting specific services in the Azure portal or anything related to the overall launch of the code.
+* .NET Core cross-platform development
```
Comment (RyanToth3) [inline README.md:21]: It would be good to add a .vsconfig to help with this setup

File: ClientServerApp/ClientApp/Program.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+                while (!cancellation.IsCancellationRequested && !string.IsNullOrEmpty(line = Console.ReadLine()))
+                {
+                    using var proxy = await client.GetProxyAsync<IEchoService>(
+                        ServiceDescriptors.EchoService, cancellationToken: cancellation.Token)
+                        .ConfigureAwait(false);
+
+                    try
+                    {
+                        var echo = await proxy.Proxy.EchoAsync(line, cancellation.Token).ConfigureAwait(false);
```
Comment (RyanToth3) [inline ClientServerApp/ClientApp/Program.cs:51]: null check here as well

**Comment by AArnott** [comment]
@kzu Do we really need a new repo for this? We already have https://github.com/microsoft/VSSDK-Extensibility-Samples that is intended for VS SDK samples.

**Comment by AArnott** [comment]
We need a .gitattributes file that specifies line ending policy, please.

File: ClientServerApp/ClientApp/Program.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                Console.WriteLine("Shutting down...");
+                cancellation.Cancel();
+            };
+
+            // Startup server app we'll communicate with. This could be in the cloud.
+            var hostStream = await StartServerAsync(serverPath);
+            var hostBroker = await RemoteServiceBroker.ConnectToMultiplexingServerAsync(hostStream);
+
+            using (var client = new ServiceBrokerClient(hostBroker))
```
Comment (AArnott) [inline ClientServerApp/ClientApp/Program.cs:38]: `ServiceBrokerClient` is intended for storing in a field, as required for the target scenario of acquiring proxies that out-live a single async method. Please just use `IServiceBroker` directly so the sample can show off intended use cases.

File: ClientServerApp/ClientApp/ClientApp.csproj [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net472</TargetFramework>
```
Comment (AArnott) [inline ClientServerApp/ClientApp/ClientApp.csproj:5]: Can you set `LangVersion=8` and `Nullable=enable` properties here?

File: ClientServerApp/Contracts/Contracts.csproj [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+        <AppendTargetFrameworkToOutputPath>true</AppendTargetFrameworkToOutputPath>
```
Comment (AArnott) [inline ClientServerApp/Contracts/Contracts.csproj:5]: Please add `LangVersion=8` and `Nullable=true` here too.
Or maybe add it to `Directory.Build.props` so it's everywhere.

File: ClientServerApp/Directory.Build.props [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
+
+    <!-- Set defaults early so .csproj can use them -->
+    <Configuration>Debug</Configuration>
+    <OutputPath>bin\$(Configuration)\</OutputPath>
+    <IntermediateOutputPath>obj\$(Configuration)\</IntermediateOutputPath>
+
+    <SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
```
Comment (AArnott) [inline ClientServerApp/Directory.Build.props:16]: IMO we should pin the SDK version anyway, in which case we'll never see the message.

File: ClientServerApp/Directory.Build.props [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<Project>
+
+  <PropertyGroup>
+    <LangVersion>preview</LangVersion>
```
Comment (AArnott) [inline ClientServerApp/Directory.Build.props:4]: This 'floats', which means what you build or open today isn't the same thing you'll get next year. I prefer to 'pin' the version.

File: ClientServerApp/Directory.Build.props [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    <!-- We generally target a single framework, so we don't need the added subdir -->
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <!-- We never emit runtime-specific assemblies, so opt-out of this. -->
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
+
+    <DefaultItemExcludes>$(DefaultItemExcludes);*.binlog;*.zip</DefaultItemExcludes>
+
+    <!-- There are some public nuget versions mismatched, but we can ignore that in the sample -->
+    <NoWarn>NU1603</NoWarn>
```
Comment (AArnott) [inline ClientServerApp/Directory.Build.props:26]: Should we fix this in our public SDK though, in order to avoid the warnings for customers?

**Comment by kzu** [comment]
I'm closing this and will delete the repo since there doesn't seem to be consensus that this is even useful to anyone. May find its way in the VSSDK samples, maybe. We'll see.

**Pull request closed without merging**